File: backend/app/fallback/llm/ollama.py
import json
import logging
import httpx

from .base import LLMProvider, PageState, AgentAction
from .json_utils import safe_json_loads as _safe_json_loads

logger = logging.getLogger(__name__)

_SYSTEM_PROMPT = (
    "You are a browser automation agent. "
    "Analyze the page and respond with a JSON object only: "
    '{"action_type":"click"|"fill"|"navigate"|"wait"|"done"|"error",'
    '"selector":"...","value":"...","url":"...","reasoning":"..."}'
)


class OllamaProvider(LLMProvider):
    """
    Local on-prem LLM provider via Ollama REST API.

    Recommended models for our use case (Taiwan HIS scraping → FHIR):
      - qwen2.5vl:7b           ← default; vision + Chinese + JSON output, ~5GB
      - qwen3:8b               ← text-only, strongest reasoning (no vision fallback)
      - qwen2.5vl:32b          ← higher quality but ~20GB, tight on 32GB RAM
      - llama3.2-vision:11b    ← alternative

    Set OLLAMA_MODEL in .env to switch models.
    """

    def __init__(self, base_url: str = "http://host.docker.internal:11434", model: str = "qwen2.5vl:7b"):
        self.base_url = base_url.rstrip("/")
        self.model = model

    @property
    def provider_name(self) -> str:
        return "ollama"

    async def decide_action(self, page_state: PageState) -> AgentAction:
        images = [page_state.screenshot_b64] if page_state.screenshot_b64 else []

        history_lines = ""
        if page_state.history:
            history_lines = "\nRecent: " + "; ".join(
                f"{a.action_type}({a.selector or a.url or ''})"
                for a in page_state.history[-3:]
            )

        prompt = (
            f"{_SYSTEM_PROMPT}\n\n"
            f"Task: {page_state.task_description}\n"
            f"URL: {page_state.url}"
            f"{history_lines}\n"
            f"HTML: {page_state.html_snippet[:2000]}\n\n"
            "JSON response:"
        )

        data = await self._generate(prompt, images)
        return self._parse_action(data.get("response", "{}"))

    async def extract_structured_data(self, html_content: str, extraction_schema: dict) -> dict:
        # Browser-captured HTML is typically 3–5× larger than the raw HTTP
        # response. 8 KB cut off the body of real HIS pages and the LLM
        # silently returned nulls. Local models are slower with bigger
        # context, so 24 KB is a compromise — bump higher if your HIS pages
        # are still being truncated.
        truncated = html_content[:24000]
        prompt = (
            f"Extract medical data from the HTML below. Match this schema exactly:\n"
            f"{json.dumps(extraction_schema, ensure_ascii=False, indent=2)}\n\n"
            f"HTML:\n{truncated}\n\n"
            "Respond with ONLY a single JSON object that conforms to the schema. "
            "No prose, no markdown fences."
        )
        data = await self._generate(prompt, [])
        raw = data.get("response", "{}")
        try:
            return _safe_json_loads(raw)
        except ValueError as exc:
            logger.error(f"Ollama returned unparseable JSON: {raw[:500]}")
            raise RuntimeError(f"Ollama returned unparseable JSON: {exc}")

    async def _generate(self, prompt: str, images: list[str]) -> dict:
        # `format=json` asks Ollama to constrain output to valid JSON — honored
        # by Qwen, Llama 3.x, and most modern local models.
        payload: dict = {
            "model": self.model,
            "prompt": prompt,
            "stream": False,
            "format": "json",
            "options": {"temperature": 0.1},
        }
        if images:
            payload["images"] = images

        async with httpx.AsyncClient(timeout=180) as client:
            resp = await client.post(f"{self.base_url}/api/generate", json=payload)
            resp.raise_for_status()
            return resp.json()

    def _parse_action(self, text: str) -> AgentAction:
        try:
            data = _safe_json_loads(text)
        except ValueError:
            return AgentAction(action_type="error", reasoning=f"Failed to parse LLM response: {text[:200]}")

        return AgentAction(
            action_type=data.get("action_type", "error"),
            selector=data.get("selector"),
            value=data.get("value"),
            url=data.get("url"),
            extracted_data=data.get("extracted_data"),
            reasoning=data.get("reasoning"),
        )

File: backend-ts/src/fallback/llm/ollama.ts
/**
 * Local Ollama LLM provider.
 *
 * Port of `backend/app/fallback/llm/ollama.py`. Uses Ollama's /api/generate
 * REST endpoint with `format=json` for JSON-constrained output. Recommended
 * models: qwen2.5vl:7b (default, vision+Chinese), qwen3:8b (text-only),
 * llama3.2-vision:11b.
 */

import type { AgentAction, LLMProvider, PageState } from "@/fallback/llm/base";
import { safeJsonLoads } from "@/fallback/llm/json-utils";

const SYSTEM_PROMPT =
  "You are a browser automation agent. " +
  "Analyze the page and respond with a JSON object only: " +
  '{"action_type":"click"|"fill"|"navigate"|"wait"|"done"|"error",' +
  '"selector":"...","value":"...","url":"...","reasoning":"..."}';

export class OllamaProvider implements LLMProvider {
  readonly providerName = "ollama";
  private baseUrl: string;
  private model: string;

  constructor(baseUrl = "http://host.docker.internal:11434", model = "qwen2.5vl:7b") {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.model = model;
  }

  async decideAction(pageState: PageState): Promise<AgentAction> {
    const images = pageState.screenshot_b64 ? [pageState.screenshot_b64] : [];

    let historyLines = "";
    if (pageState.history && pageState.history.length > 0) {
      historyLines = `\nRecent: ${pageState.history
        .slice(-3)
        .map((a) => `${a.action_type}(${a.selector ?? a.url ?? ""})`)
        .join("; ")}`;
    }

    const prompt = `${SYSTEM_PROMPT}\n\nTask: ${pageState.task_description}\nURL: ${pageState.url}${historyLines}\nHTML: ${pageState.html_snippet.slice(0, 2000)}\n\nJSON response:`;

    const data = await this.generate(prompt, images);
    const text = typeof data.response === "string" ? data.response : "{}";
    let parsed: any;
    try {
      parsed = safeJsonLoads(text);
    } catch {
      return {
        action_type: "error",
        reasoning: `Failed to parse LLM response: ${text.slice(0, 200)}`,
      };
    }
    return {
      action_type: parsed.action_type ?? "error",
      selector: parsed.selector ?? null,
      value: parsed.value ?? null,
      url: parsed.url ?? null,
      extracted_data: parsed.extracted_data ?? null,
      reasoning: parsed.reasoning ?? null,
    };
  }

  async extractStructuredData(
    htmlContent: string,
    extractionSchema: Record<string, any>,
  ): Promise<any> {
    const truncated = htmlContent.slice(0, 24000);
    const prompt = `Extract medical data from the HTML below. Match this schema exactly:\n${JSON.stringify(extractionSchema, null, 2)}\n\nHTML:\n${truncated}\n\nRespond with ONLY a single JSON object that conforms to the schema. No prose, no markdown fences.`;
    const data = await this.generate(prompt, []);
    const raw = typeof data.response === "string" ? data.response : "{}";
    try {
      return safeJsonLoads(raw);
    } catch (e: any) {
      throw new Error(`Ollama returned unparseable JSON: ${e?.message ?? e}`);
    }
  }

  private async generate(prompt: string, images: string[]): Promise<any> {
    const payload: Record<string, any> = {
      model: this.model,
      prompt,
      stream: false,
      format: "json",
      options: { temperature: 0.1 },
    };
    if (images.length > 0) payload.images = images;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 180_000);
    try {
      const resp = await fetch(`${this.baseUrl}/api/generate`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
        signal: controller.signal,
      });
      if (!resp.ok) {
        const errText = await resp.text();
        throw new Error(`Ollama HTTP ${resp.status}: ${errText.slice(0, 200)}`);
      }
      return await resp.json();
    } finally {
      clearTimeout(timer);
    }
  }
}
